src/Xrefcheck/Config.hs

```diff
@@ -145,6 +228,32 @@ instance FromJSON (NetworkingConfig' Maybe) where
 instance FromJSON (NetworkingConfig) where
   parseJSON = genericParseJSON aesonConfigOption
 
+instance FromJSON (RedirectRule) where
+  parseJSON = genericParseJSON aesonConfigOption
+
+instance FromJSON (RedirectRuleOutcome) where
+  parseJSON = withText "Redirect rule outcome" $
+    \case
+      "valid" -> pure RROValid
+      "invalid" -> pure RROInvalid
+      "follow" -> pure RROFollow
+      _ -> fail "expected (valid|invalid|follow)"
+
+instance FromJSON (RedirectRuleOn) where
+  parseJSON v = code v
+    <|> text v
+    <|> fail "expected a redirect (3XX) HTTP code or (permanent|temporary)"
+    where
+      code cv = do
+        i <- parseJSON @Int cv
+        unless (isRedirectCode i) $ fail mempty
```

Comment:
I set the error message as empty in both auxiliary parsers (`code` and `text`) because they are used as follows:

```
code <|> text <|> fail "A common error message"
```

Is there a better way to combine both error messages?